### PR TITLE
ENH: implement lazy imports for optional dependencies

### LINKED ISSUE #27: Optimize importing unyt
As reported by @majj, importing unyt takes about one second. I suspect there isn't much we can do about this since we need to import both numpy and sympy.

### PR DESCRIPTION
**Pull request by neutrinoceros**
This is another step to resolve #27
Overall, this gets us to a state where core dependencies (numpy + sympy) are responsible for at least 50-60 % of unyt's import time.

This refactor is similar to my work in yt https://github.com/yt-project/yt/pull/3935
It allows for optional dependencies to never be imported before they are actually requested by users, reducing unyt's own startup time when heavy optional dependencies are installed.

Making matplotlib a true "on-demand" import took a little more work, but now `import unyt` doesn't cause `matplotlib` to be imported too when available, reducing the startup time by about 15% to 25% with respect to unyt 2.9.2 (on my machine).

### CODE REVIEW

**Comment by neutrinoceros** [comment]
this now includes changes for dask, recently added with #185 (ping @chrishavlin)

**Comment by neutrinoceros** [comment]
this is now based on #272

**Comment by neutrinoceros** [comment]
@jzuhone @chrishavlin can I request your reviews on this ?

File: tox.ini [hunk 1]
```diff
@@ -57,9 +57,9 @@ deps =
     dask[array,diagnostics]>=2021.04.1
 depends = begin
 commands =
-    # don't do doctests in rst files due to lack of way to specify optional
```
Comment (jzuhone) [inline tox.ini:60]: Is there a reason you took this out here? It wasn't mentioned in the description and it seems unrelated to load times.

**Review by jzuhone — APPROVED** [review]
Curious about the doctests bit, but otherwise looks good.

File: tox.ini (hunk 1 above)
Comment (neutrinoceros) [reply tox.ini:60]: I can't recall why I did that. Let's try to revert this part and see if anything fails.

Comment (neutrinoceros) [reply tox.ini]: welp, nothing breaks. I'll just clean up the history

**Review by chrishavlin — APPROVED** [review]
Looks good! Left a couple minor comments.

One additional question -- why'd you put the `unyt_arrayConverter` class in `_mpl_array_converter.__init__.py` instead of a new module, `_mpl_array_converter.py` ?

File: unyt/_on_demand_imports.py [hunk 2]
```diff
@@ -31,182 +34,142 @@ def __getattr__(self, item):
...
-    _log = None
-    _units = None
-    _version = None
+    def __init_subclass__(cls):
+        if not cls.__name__.endswith("_imports"):
+            raise TypeError(f"class {cls}'s name needs to be suffixed '_imports'")
+
+    def __new__(cls):
+        if cls is OnDemand:
+            raise TypeError("The OnDemand base class cannot be instanciated.")
```
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply unyt/_on_demand_imports.py:50]: good catch !

File: unyt/_on_demand_imports.py [hunk 3]
```diff
@@ -31,182 +34,142 @@ def __getattr__(self, item):
...
-                from matplotlib import pyplot
-            except ImportError:
-                pyplot = NotAModule(self._name)
-            self._pyplot = pyplot
-        return self._pyplot
+        from matplotlib import pyplot
+
+        return pyplot
 
     @property
```
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply unyt/_on_demand_imports.py:145]: not intentional, thanks ! (same with `use`)

**Comment by neutrinoceros** [comment]
> One additional question -- why'd you put the unyt_arrayConverter class in _mpl_array_converter.__init__.py instead of a new module, _mpl_array_converter.py ?

I can't remember. I would guess I went through several iterations and that form was the first one with which I managed to pass tests, for some reason. Now I agree that a module is better than a mono-module package, and I've checked that it does not in fact break anything, so let's roll with that.
